contracts/v2/core/WithdrawQueue.sol

```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
+import {Pausable} from "@openzeppelin/contracts/security/Pausable.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+
+contract WithdrawQueue is AccessControl, Pausable, ReentrancyGuard {
+    using Address for address payable;
+
+    struct Queue {
+        address to;
+        uint256 amount;
+    }
+
+    uint256 public constant CHUNK_SIZE = 32 ether;
+
+    uint256 front;
+    uint256 end;
+    uint256 lockedFront;
+    uint256 lockedAmount;
+
+    mapping(uint256 => Queue) public queue;
+
+    error NoAvailableQueue();
+
+    constructor() {
+        grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    function push(address to) external payable nonReentrant {
+        Queue memory newWithdraw = Queue({to: to, amount: msg.value});
```

Comment:
push and process should handle token transfers too. 

wsgeth/sgeth input for push, held in contract, validated as correct token, like the pre-existing withdrawals contract 
sgeth/wsgeth burn on processWithdraw perhaps to make tracking easier. 
also eth needs to be sent out back to the user, and total sent needs to be checked / tracked

Reply:
how can I convert wsgEth to eth value?

Reply:
price per share should do it
https://github.com/chimera-defi/SharedDeposit/blob/main/test/v2/core/wsgETH.spec.js#L125

https://github.com/chimera-defi/SharedDeposit/blob/main/contracts/v2/core/WSGEth.sol#L72

should be tested

Reply:
we should track total eth value on queue to make chunk withdraw functionality, and wsgeth/eth rate can be changed by time(when push and process)

Reply:
yea we want to pick 1 price, ideally the price when the user deposits, and use that throughout the lifecycle of that withdrawal request. 
might be other edge cases were missing, but i think thats what other protocols do